src/main/scala/rise/core/DSL/infer.scala

```diff
@@ -27,21 +28,20 @@ object infer {
   }
 
   def printTypesAndTypeHoles(expr: Expr): Unit = {
-    // TODO: move holeFound state into the traverse
-    var holeFound = false
-    traverse(expr, new PureExprTraversal {
-      override def expr : Expr => Pure[Expr] = {
+    val T = new PairMonoidTraversal[Boolean, Pure] {
+      override val fstMonoid = OrMonoid
+      override val wrapperMonad = PureMonad
+      override def expr: Expr => Pair[Expr] = {
         case h@primitives.typeHole(msg) =>
           println(s"found type hole ${msg}: ${h.t}")
-          holeFound = true
-          return_(h : Expr)
+          Pure((true, h))
         case p@primitives.printType(msg) =>
           println(s"$msg : ${p.t} (Rise level)")
-          return_(p : Expr)
+          return_(p: Expr)
         case e => super.expr(e)
       }
-    })
-    if (holeFound) {
+    }
+    if (T.expr(expr).unwrap._1) {
```

Comment:
I find mutability to work quite well for such "global accumulation" patterns. The monoidal version is more noisy and the code is harder to read for me.

Reply:
The trouble with that approach is that you cannot easily reuse the traversal (e.g. to use it once on an expression and once on a type). I have slightly reduced the noise by getting rid of the `unwrap`s and adding a `record` function so it is slightly more semantically meaningful.

Reply:
With the mutable approach, can you not just accept the mutated accumulator as a parameter to the traversal/visitor to make it reusable? And to avoid surrounding the traversal with the environment on use can you not have a default accumulator and return it? like:
```scala
def traverseAndAccumulate(e: Expr, acc: mutable.HashMap = HashMap()): mutable.HashMap = {
  traverse(...) // acc += stuff
  acc
}
```
Anyway, we can also go for the Monad way, I just find it less straight-forward.